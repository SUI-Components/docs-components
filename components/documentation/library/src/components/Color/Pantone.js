/* eslint react/prop-types: 0 */
import React, {forwardRef, useState, useEffect, useMemo} from 'react'
import cx from 'classnames'
import debounce from 'lodash.debounce'
import {useCopyToClipboard} from 'react-use'
import Color from 'color'

import useControlledState from '../../hooks/useControlledState/useControlledState'
import useColorAPINameHook from './useColorAPINameHook.hook'
import Base from '../Base.core'

import './Pantone.scss'

const fromPercent = valNum => {
  const decimalValue = Math.round((valNum * 255) / 100)
  let hexValue
  if (valNum < 7) {
    hexValue = '0' + decimalValue.toString(16).toUpperCase()
  } else {
    hexValue = decimalValue.toString(16).toUpperCase()
  }
  return hexValue
}

const DISPLAY_MODES = {
  HEX: color => `${color.hex()}${fromPercent(color.valpha * 100)}`,
  RGB: color => color.rgb().string(),
  HSL: color => color.hsl().string(),
  CMYK: color =>
    `cmyk(${color
      .cmyk()
      .round()
      .array()
      .join(', ')
      .toString()})`
}

const DEFAULT_DISPLAY_MODE = 'HEX'
const DEFAULT_COLOR = 'transparent'

const clickHandler = ({
  color,
  setMode,
  mode,
  onClick,
  onDoubleClick,
  tokenName,
  copyToClipboard
}) => event => {
  switch (event.detail) {
    case 1:
      if (mode || tokenName !== undefined) {
        copyToClipboard(tokenName || DISPLAY_MODES[mode](color))
      }
      if (typeof onClick === 'function') {
        onClick(event, color, mode)
      }
      break
    case 2:
    default:
      setMode(Pantone.next({mode, color}))
      if (typeof onDoubleClick === 'function') {
        onDoubleClick(event, color, mode)
      }
      break
  }
}

/**
 * Pantone card of the background element color
 */
const Pantone = forwardRef(
  (
    {
      elementType = 'div',
      color: colorProp = DEFAULT_COLOR,
      displayMode,
      initialDisplayMode = DEFAULT_DISPLAY_MODE,
      className,
      onClick,
      onDoubleClick,
      onDataFetched,
      name,
      tokenName,
      ...props
    },
    forwardedRef
  ) => {
    const [mode, setMode] = useControlledState(displayMode, initialDisplayMode)
    const [data, setData] = useState({
      value: name !== undefined ? name : undefined
    })
    const [color, setColor] = useState(Color(colorProp))
    const [, copyToClipboard] = useCopyToClipboard()
    useColorAPINameHook(
      {HEXColor: DISPLAY_MODES.HEX(color).slice(1, 7), name: data.value},
      value => {
        name === undefined && setData(value)
        typeof onDataFetched === 'function' && onDataFetched(value)
      }
    )
    useEffect(() => {
      setData({value: name !== undefined ? name : undefined})
    }, [name, setData])
    useEffect(() => {
      const value = Color(colorProp)
      setColor(value)
    }, [colorProp, setColor])
    const handleClick = useMemo(
      () =>
        debounce(
          clickHandler({
            color,
            setMode,
            mode,
            onClick,
            onDoubleClick,
            tokenName,
            copyToClipboard
          }),
          200
        ),
      [color, setMode, mode, onClick, onDoubleClick, tokenName]
    )
    const tiltColor = useMemo(() => {
      if (typeof color !== 'object') {
        return 'transparent'
      } else if (color.isLight()) {
        return DISPLAY_MODES.HEX(color.darken(0.3))
      } else if (color.isDark()) {
        return DISPLAY_MODES.HEX(color.lighten(0.3))
      }
    }, [color])

    return (
      <Base
        {...props}
        elementType={elementType}
        className={cx('sui-studio-doc', 'sui-studio-doc-pantone', className)}
        role="button"
        onClick={handleClick}
        style={{
          backgroundColor: color ? DISPLAY_MODES[mode](color) : DEFAULT_COLOR
        }}
        ref={forwardedRef}
      >
        <div className="sui-studio-doc-pantone-frame-wrapper">
          <Base
            elementType="div"
            className={cx('sui-studio-doc', 'sui-studio-doc-pantone-frame', {
              'sui-studio-doc-pantone-frame-isLight':
                color?.isLight && color.isLight(),
              'sui-studio-doc-pantone-frame-isDark':
                color?.isDark && color.isDark()
            })}
          >
            <Base
              elementType="div"
              className={cx(
                'sui-studio-doc',
                'sui-studio-doc-pantone-token-name',
                {
                  'sui-studio-doc-pantone-token-name-isLight':
                    color?.isLight && color.isLight(),
                  'sui-studio-doc-pantone-token-name-isDark':
                    color?.isDark && color.isDark()
                }
              )}
            >
              <span className="sui-studio-doc-pantone-text">{tokenName}</span>
            </Base>
            <Base
              elementType="div"
              className={cx('sui-studio-doc', 'sui-studio-doc-pantone-legend', {
                'sui-studio-doc-pantone-legend-isLight':
                  color?.isLight && color.isLight(),
                'sui-studio-doc-pantone-legend-isDark':
                  color?.isDark && color.isDark()
              })}
            >
              <span className="sui-studio-doc-pantone-text">
                {typeof mode === 'string' && DISPLAY_MODES[mode](color)}
              </span>
            </Base>
            <Base
              elementType="div"
              fontWeight="bold"
              className={cx('sui-studio-doc', 'sui-studio-doc-pantone-name', {
                'sui-studio-doc-pantone-legend-isLight':
                  color?.isLight && color.isLight(),
                'sui-studio-doc-pantone-legend-isDark':
                  color?.isDark && color.isDark()
              })}
            >
              <span className="sui-studio-doc-pantone-text">
                {data.value !== null && data.value !== undefined
                  ? data.value
                  : 'â€Ž'}
              </span>
            </Base>
            <span
              className="sui-studio-doc-pantone-frame-tilt"
              style={{backgroundColor: tiltColor}}
            />
            <span
              className="sui-studio-doc-pantone-frame-tilt"
              style={{backgroundColor: tiltColor}}
            />
            <span
              className="sui-studio-doc-pantone-frame-tilt"
              style={{backgroundColor: tiltColor}}
            />
            <span
              className="sui-studio-doc-pantone-frame-tilt"
              style={{backgroundColor: tiltColor}}
            />
          </Base>
        </div>
      </Base>
    )
  }
)
Pantone.displayName = 'Pantone'
Pantone.propTypes = {}
Pantone.defaultProps = {}
Pantone.mode = ({color, mode}) => {
  const modeKeys = Object.keys(DISPLAY_MODES)
  if (modeKeys.includes(mode)) {
    return mode
  }
  let response
  switch (color.model) {
    case 'hsl':
      response = modeKeys.find(value => value === 'HSL')
      break
    case 'rgb':
    default:
      response = modeKeys.find(value => value === 'RGB')
      break
  }
  return response
}
Pantone.next = ({mode, color}) => {
  const modeKeys = Object.keys(DISPLAY_MODES)
  const index = modeKeys.indexOf(Pantone.mode({mode, color}))
  if (index >= 0) {
    return modeKeys[(index + 1) % modeKeys.length]
  } else {
    return 'RGB'
  }
}
Pantone.displayMode = mode => color => DISPLAY_MODES[mode](color)

export default Pantone

export {Pantone}
