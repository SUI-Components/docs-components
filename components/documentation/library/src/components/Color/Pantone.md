# Pantone
Pantone card of the background element color

## Installation
`npm install @s-ui/documentation-library`

## Usage

```js
// import Pantone DOC component
import { Pantone } from '@s-ui/documentation-library/lib/components/Pantone/Pantone.js'
```

or

```js
// import all DOC components you need
import { Pantone } from '@s-ui/documentation-library'
```

## Prop-Types

| Name | Description | Values  | Default | Required |
| ---: |:---:| ---:| ---: |:---: |
| elementType | undefined | | |  [ ]  |
| color | undefined | | |  [ ]  |
| initialDisplayMode | undefined | | |  [ ]  |
