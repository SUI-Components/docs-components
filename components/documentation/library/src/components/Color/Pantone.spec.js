/* global expect */
import React, {createRef} from 'react'
import ReactDOM from 'react-dom'

import {Pantone} from './Pantone'

import setupEnvironment from '../../helpers/setupEnvironment'

describe('components', () => {
  describe('Pantone', () => {
    describe('default', () => {
      const Component = Pantone
      const setup = setupEnvironment(Component)

      test('should render without crashing', () => {
        // Given
        const props = {style: {backgroundColor: 'red'}, name: null}

        // When
        const component = (
          <Component {...props}>
            {({color}) => {
              return color.hex()
            }}
          </Component>
        )

        // Then
        const div = document.createElement('div')
        ReactDOM.render(component, div)
        ReactDOM.unmountComponentAtNode(div)
      })

      test('should NOT render null', async () => {
        // Given
        const props = {
          color: 'red'
        }
        const label = 'Red'

        // When
        const {container, findByText, getByText} = setup(props)

        // Then
        expect(container.innerHTML).toBeString()
        expect(container.innerHTML).not.toHaveLength(0)

        // color name label
        await findByText(label)
        const nameLabelElement = getByText(label)

        expect(nameLabelElement.innerHTML).toBe(label)
      })

      test('should have propTypes and defaultProps ', () => {
        // Given
        const component = Component

        // When

        // Then
        expect(component.hasOwnProperty('propTypes')).toBeTruthy()
        expect(component.hasOwnProperty('defaultProps')).toBeTruthy()
      })

      test('should have proper displayName attribute', () => {
        // Given
        // const props = {color: 'red'}

        // When
        const component = Component

        // Then
        expect(component.displayName).toBeString()
        expect(component.displayName).not.toHaveLength(0)
        expect(component.displayName).toMatchSnapshot()
      })

      describe('forwardRef', () => {
        test('should have reference', () => {
          // Given
          const props = {
            color: 'red',
            name: null
          }
          const ref = createRef()

          // When
          const component = <Component {...props} ref={ref} />
          const div = document.createElement('div')
          ReactDOM.render(component, div)

          // Then
          expect(ref.current).not.toBe(undefined)
          expect(ref.current.nodeName).toBe('DIV')
        })

        test('should have defined elementType reference', () => {
          // Given
          const props = {
            elementType: 'span',
            color: 'red',
            name: null
          }
          const ref = createRef()

          // When
          const component = <Component {...props} ref={ref} />
          const div = document.createElement('div')
          ReactDOM.render(component, div)

          // Then
          expect(ref.current).not.toBe(undefined)
          expect(ref.current.nodeName).toBe('SPAN')
        })
      })
    })
  })
})
