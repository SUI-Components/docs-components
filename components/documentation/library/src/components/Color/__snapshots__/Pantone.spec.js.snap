// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Pantone default should have proper displayName attribute 1`] = `"Pantone"`;
